File: CarRentalSystem/CarRentalSystemController/CarRentalCommands.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace sf.systems.rentals.cars
{
    public static class CarRentalCommands
    {
        public static CarRentalContext ReturnCar(CarRentalContext rentalContext, CarRentalContextAction posteriorReturnCar)
        {
            // validation
            validateContext(rentalContext);

            // go!
            rentalContext.RentalTransaction = rentalContext.CarRentalSystem.ReturnCar(
                rentalContext.CarRentalSystem.LookupCustomer(rentalContext.CustomerID));

            // finally
            rentalContext.ActionCompleted = rentalContext.RentalTransaction.IsClosed;
            
            if (rentalContext.ActionCompleted)
                rentalContext = posteriorReturnCar?.Invoke(rentalContext);
            
            return rentalContext;
        }

        public static CarRentalContext RentCar(CarRentalContext rentalContext, 
            CarRentalContextAction notifierEmptyCustomer,
            CarRentalContextAction notifierCarIsnotAvaliable,
            CarRentalContextAction posteriorReturnCar
            )
        {
            // validation
            validateContext(rentalContext);

            // go!
            // validate params
            var validationError = validateCustomer(rentalContext, notifierEmptyCustomer);
            var car = rentalContext.CarRentalSystem.GetCar(rentalContext.CarID);
            validationError = validateCar(rentalContext, notifierCarIsnotAvaliable, car);
            
            // exit if a validaton is occured
            if (validationError) 
                return rentalContext;

            // Rent fisrt avaliable car by given customer for 3 days
            rentalContext.RentalTransaction = rentalContext.CarRentalSystem.RentCar(
                rentalContext.CustomerID,
                rentalContext.CarID,
                rentalContext.RentFrom,
                rentalContext.RentTo
                );

            rentalContext.ActionCompleted = 
                rentalContext.RentalTransaction.Id != null 
                && rentalContext.RentalTransaction.Id.Trim().Length > 1;

            if (rentalContext.ActionCompleted)
                rentalContext = posteriorReturnCar?.Invoke(rentalContext);

            return rentalContext;
        }

        public static void RegisterCustomer(CarRentalContext context, string[] argCustomer)
        {
            // init
            var rentalSystem = context.CarRentalSystem;
            var customerID = argCustomer[0];

            // go
            var customer = rentalSystem.LookupCustomer(customerID);
            if (customer == null)
            {
                customer = rentalSystem.RegisterCustomer(argCustomer[0], argCustomer[1], argCustomer[2], argCustomer[3], argCustomer[4]);
                context.CarRentalSystem.LogAndShowMessage(
                    $"New Customer \"{customer.Name}\" (ID: {customer.Id}) has been successfully added to the System!");
            }
            else
            {
                context.CarRentalSystem.LogAndShowMessage(
                    $"Customer \"{customer.Name}\" (ID: {customer.Id}) is already registered in the System!");
            }
        }

        public static void DeleteCustomer(CarRentalContext context, string customerId)
        {
            // init
            var rentalSystem = context.CarRentalSystem;
            var customer = rentalSystem.LookupCustomer(customerId);

            // go
            if (rentalSystem.DeleteCustomer(customerId))
            {
                context.CarRentalSystem.LogAndShowMessage(
                    $"Customer \"{customer.Name}\" (ID: {customer.Id}) has been successfully deleted from the System!");
            }
            else
            {
                context.CarRentalSystem.LogAndShowMessage(
                    $"Customer with ID: {customerId} has not been found!");
            }
        }

        public static void DeleteCar(CarRentalContext context, string carId)
        {
            if (context.CarRentalSystem.DeleteCar(carId))
            {
                context.CarRentalSystem.LogAndShowMessage(
                    $"Car with ID: {carId} has been successfully deleted from the System!");
            }
        }

        public static void AddCar(CarRentalContext context, string[] argCar)
        {
            // init
            var carRentalSystem = context.CarRentalSystem;
            var carId = argCar[0];
            var car = carRentalSystem.LookupCar(carId);

            // go!
            if (car == null)
            {
                var newCar = context.CarRentalSystem.AddCar(argCar[0], argCar[1], argCar[2], Convert.ToInt32(argCar[3]), Convert.ToDouble(argCar[4]));
                context.CarRentalSystem.LogAndShowMessage($"New Car with ID: {newCar.Id} has been successfully added to the System!");
            }
            else
                context.CarRentalSystem.LogAndShowMessage($"Car with ID: {car.Id} already exists in the System!");
        }

        public static void SaveData(CarRentalContext context)
        {
            // Save the data to disk
            context.CarRentalSystem.LogAndShowMessage("\nSaving data ...");
            context.CarRentalSystem.SaveData();
            context.CarRentalSystem.LogAndShowMessage("All data has been saved.\n");
        }

        public static void LoadData(CarRentalContext context)
        {
            // Load the data from disk
            context.CarRentalSystem.LogAndShowMessage("\nLoading data ...");
            context.CarRentalSystem.LoadData();
            context.CarRentalSystem.LogAndShowMessage("All data has been loaded.\n");
        }

        public static CarRentalContext ReturnCar(CarRentalContext context)
        {
            return ReturnCar(context,
                    (CarRentalContext ctx) =>
                    {
                        context.CarRentalSystem.LogAndShowMessage(
                            $"\nCustomer rental is closed (ID-Customer:{context.CustomerID}, ID-Rental: {context.RentalTransaction.Id}).\n");

                        return ctx;
                    }
                );
        }

        public static CarRentalContext RentCar(CarRentalContext context)
        {
            return RentCar(context,
                    // empty customer
                    (CarRentalContext ctx) =>
                    {
                        ctx.CarRentalSystem.LogAndShowMessage("Customer-ID is empty!");
                        return ctx;
                    },
                    // car is not avaliable
                    (CarRentalContext ctx) =>
                    {
                        ctx.CarRentalSystem.LogAndShowMessage($"Car with ID:{ctx.CarID} is not avaliable!");
                        return ctx;
                    },
                    // posterior - rent message
                    (CarRentalContext ctx) =>
                    {
                        ctx.CarRentalSystem.LogAndShowMessage(
                            $"\nCar with (ID-Car:{ctx.CarID}) has been rented by Customer (ID-Customer:{ctx.CustomerID}," +
                            $" ID-Rental: {ctx.RentalTransaction.Id}).\n");
                        return ctx;
                    }
                );
        }
        
        private static void validateContext(CarRentalContext rentalContext)
        {
            if (rentalContext == null) throw new ArgumentNullException("rentalContext");
            if (rentalContext.CarRentalSystem == null) throw new ArgumentNullException("rentalContext.carRentalSystem");
        }
        
        private static bool validateCar(CarRentalContext rentalContext, CarRentalContextAction notifierCarIsnotAvaliable, object car)
        {
            var validationError = false;

            if (car == null)
            {
                notifierCarIsnotAvaliable?.Invoke(rentalContext);
                validationError = true;
            }

            return validationError;
        }

        private static bool validateCustomer(CarRentalContext rentalContext, CarRentalContextAction notifierEmptyCustomer)
        {
            var validationError = false;

            if (string.IsNullOrEmpty(rentalContext.CustomerID))
            {
                notifierEmptyCustomer?.Invoke(rentalContext);
                validationError = true;
            }

            return validationError;
        }
    }
}

File: CarRentalSystemPython/controller/CarRentalCommands.py
import datetime
from typing import List, Union, Callable
from typing import Callable, Union
from controller.CarRentalContext import CarRentalContext
from controller.CarRentalContextAction import CarRentalContextAction


class CarRentalCommands:
    @staticmethod
    def return_car(rentalContext: CarRentalContext, posteriorReturnCar: CarRentalContextAction) -> CarRentalContext:
        # validation
        CarRentalCommands.validate_context(rentalContext)

        # go!
        rentalContext.RentalTransaction = rentalContext.car_rental_system.return_car(
            rentalContext.car_rental_system.lookup_customer(rentalContext.customer_id))

        # finally
        rentalContext.ActionCompleted = rentalContext.RentalTransaction.is_closed
        
        if rentalContext.ActionCompleted:
            rentalContext = posteriorReturnCar(rentalContext)
        
        return rentalContext

    @staticmethod
    def rent_car(rentalContext: CarRentalContext, 
                notifierEmptyCustomer: CarRentalContextAction, 
                notifierCarIsnotAvaliable: CarRentalContextAction, 
                posteriorReturnCar: CarRentalContextAction) -> CarRentalContext:
        # validation
        CarRentalCommands.validate_context(rentalContext)

        # go!
        # validate params
        validationError = CarRentalCommands.validate_customer(rentalContext, notifierEmptyCustomer)
        car = rentalContext.car_rental_system.get_car(rentalContext.car_id)
        validationError = CarRentalCommands.validate_car(rentalContext, notifierCarIsnotAvaliable, car)
        
        # exit if a validation is occurred
        if validationError: 
            return rentalContext

        # Rent first available car by given customer for 3 days
        rentalContext.rental_transaction = rentalContext.car_rental_system.rent_car(
            rentalContext.customer_id,
            rentalContext.car_id,
            rentalContext.rent_from,
            rentalContext.rent_to)

        rentalContext.action_completed = bool(rentalContext.rental_transaction.id)

        if rentalContext.action_completed: 
            posteriorReturnCar(rentalContext)

        return rentalContext

    @staticmethod
    def register_customer(context: CarRentalContext, argCustomer: List[str]):
        # init
        rentalSystem = context.car_rental_system
        customerID = argCustomer[0]

        # go
        customer = rentalSystem.lookup_customer(customerID)
        if customer is None:
            customer = rentalSystem.register_customer(argCustomer[0], argCustomer[1], argCustomer[2], argCustomer[3], argCustomer[4])
            rentalSystem.log_and_show_message(f'New Customer "{customer.name}" (ID: {customer.id}) has been successfully added to the System!')
        else:
            rentalSystem.log_and_show_message(f'Customer "{customer.name}" (ID: {customer.id}) is already registered in the System!')

    @staticmethod
    def delete_customer(context: CarRentalContext, customerId: str):
        # init
        rentalSystem = context.car_rental_system
        customer = rentalSystem.lookup_customer(customerId)

        # go
        if rentalSystem.delete_customer(customerId):
            rentalSystem.log_and_show_message(f'Customer "{customer.name}" (ID: {customer.id}) has been successfully deleted from the System!')
        else:
            rentalSystem.log_and_show_message(f'Customer with ID: {customerId} has not been found!')

    @staticmethod
    def delete_car(context: CarRentalContext, car_id: str) -> None:
        if context.car_rental_system.delete_car(car_id):
            context.car_rental_system.log_and_show_message(f"Car with ID: {car_id} has been successfully deleted from the system!")

    @staticmethod
    def add_car(context: CarRentalContext, arg_car: List[str]) -> None:
        car_id = arg_car[0]
        car = context.car_rental_system.lookup_car(car_id)
        
        if car is None:
            new_car = context.car_rental_system.add_car(car_id, arg_car[1], arg_car[2], int(arg_car[3]), float(arg_car[4]))
            context.car_rental_system.log_and_show_message(f"New Car with ID: {new_car.id} has been successfully added to the system!")
        else:
            context.car_rental_system.log_and_show_message(f"Car with ID: {car.id} already exists in the system!")

    @staticmethod
    def save_data(context: CarRentalContext) -> None:
        context.car_rental_system.log_and_show_message("\nSaving data ...")
        context.car_rental_system.save_data()
        context.car_rental_system.log_and_show_message("All data has been saved.\n")

    @staticmethod
    def load_data(context: CarRentalContext) -> None:
        context.car_rental_system.log_and_show_message("\nLoading data ...")
        context.car_rental_system.load_data()
        context.car_rental_system.log_and_show_message("All data has been loaded.\n")

    @staticmethod
    def validate_context(rental_context: CarRentalContext) -> None:
        if rental_context is None:
            raise ValueError("rental_context must not be None")
        if rental_context.car_rental_system is None:
            raise ValueError("rental_context.car_rental_system must not be None")
        
    @staticmethod
    def validate_car(rental_context: CarRentalContext, notifier_car_is_not_available: CarRentalContextAction, car) -> bool:
        validation_error = False
        if car is None:
            if notifier_car_is_not_available:
                notifier_car_is_not_available(rental_context)
            validation_error = True
        return validation_error
    
    @staticmethod
    def validate_customer(rental_context: CarRentalContext, notifier_empty_customer: CarRentalContextAction) -> bool:
        validation_error = False
        if not rental_context.customer_id:
            if notifier_empty_customer:
                notifier_empty_customer(rental_context)
            validation_error = True
        return validation_error
